File: src/npmpc/mpc/problem/MPCBase.py
import casadi as ca

from npmpc.dynamics.integrators import IntegrationBase

class MPCBase:
    params: dict
    integrator: IntegrationBase
    p: ca.DM
    slack_weight = 1000
    
    def cost_function(self, x: ca.MX, u: ca.MX) -> ca.MX:
        """MPC cost; subclasses read cost weights from self.params['cost']."""
        raise NotImplementedError


    def integration_constraints(self, x: ca.MX, u: ca.MX, dt: float) -> list:
        """Dynamics integration constraints over the horizon."""
        return self.integrator.casadi_implicit(x, u, dt, self.p)


    def slack_constraints(self, slack: ca.MX, slack_bound: dict) -> list:
        """Bounds on the slack variables."""
        constraints = []; x_bound = slack_bound['x']
        for i, bound in enumerate(x_bound):
            if bound[0] != ca.inf: 
                constraints.append(slack[i] >= 0)
                constraints.append(slack[i] <= bound[0])

        return constraints
    
    
    def slack_cost(self, slack: ca.MX, slack_bound: dict) -> ca.MX:
        """Penalty cost on the slack variables."""
        cost = 0; x_bound = slack_bound['x']
        for i, bound in enumerate(x_bound):
            if bound[0] != ca.inf:
                cost += self.slack_weight * 0.5 * (slack[i]**2 + slack[i])

        return cost
        

    def state_constraints(self, x: ca.MX, slack: ca.MX, params: dict) -> list[ca.MX]:
        """State bound constraints, softened by slack where allowed."""
        constraints = []
        x_bound = params['hard_bound']['x']; slack_bound = params['slack_bound']['x']
        
        for i in range(x.shape[1]):
            if x_bound[i][0] != -ca.inf:
                if slack_bound[i] != ca.inf:
                    constraints.append(x[:,i] >= x_bound[i][0] - slack[i])
                else:
                    constraints.append(x[:,i] >= x_bound[i][0])
            if x_bound[i][1] != ca.inf:
                if slack_bound[i] != ca.inf:
                    constraints.append(x[:,i] <= x_bound[i][1] + slack[i])
                else:
                    constraints.append(x[:,i] <= x_bound[i][1])
        return constraints
    
    
    def input_constraints(self, u: ca.MX, params: dict) -> list[ca.MX]:
        """Input bound constraints."""
        constraints = []; u_bound = params['hard_bound']['u']
        for i in range(u.shape[1]):
            if u_bound[i][0] != -ca.inf: 
                constraints.append(u[:,i] >= u_bound[i][0])
            if u_bound[i][1] != ca.inf: 
                constraints.append(u[:,i] <= u_bound[i][1])
        return constraints

File: src_cpp/npmpc/mpc/problem/MPCBase.cpp
#include "npmpc/mpc/problem/MPCBase.hpp"

#include <limits>

namespace npmpc::mpc::problem {

namespace {
constexpr double kInf = std::numeric_limits<double>::infinity();
}

std::vector<casadi::MX> MPCBase::slackConstraints(const casadi::MX& slack,
                                                    const std::vector<double>& slackBoundX) const {
    std::vector<casadi::MX> constraints;
    for (size_t i = 0; i < slackBoundX.size(); ++i) {
        if (slackBoundX[i] != kInf) {
            casadi::MX s = slack(0, static_cast<casadi_int>(i));
            constraints.push_back(s >= 0);
            constraints.push_back(s <= slackBoundX[i]);
        }
    }
    return constraints;
}

casadi::MX MPCBase::slackCost(const casadi::MX& slack, const std::vector<double>& slackBoundX) const {
    casadi::MX cost = 0;
    for (size_t i = 0; i < slackBoundX.size(); ++i) {
        if (slackBoundX[i] != kInf) {
            casadi::MX s = slack(0, static_cast<casadi_int>(i));
            cost = cost + slackWeight_ * 0.5 * (casadi::MX::pow(s, 2) + s);
        }
    }
    return cost;
}

std::vector<casadi::MX> MPCBase::stateConstraints(const casadi::MX& x, const casadi::MX& slack,
                                                    const HardBound& hardBound,
                                                    const std::vector<double>& slackBoundX) const {
    std::vector<casadi::MX> constraints;
    for (size_t i = 0; i < hardBound.x.size(); ++i) {
        const Bound& bound = hardBound.x[i];
        casadi::MX xi = x(casadi::Slice(), static_cast<casadi_int>(i));
        if (bound.lower != -kInf) {
            if (slackBoundX[i] != kInf) {
                constraints.push_back(xi >= bound.lower - slack(0, static_cast<casadi_int>(i)));
            } else {
                constraints.push_back(xi >= bound.lower);
            }
        }
        if (bound.upper != kInf) {
            if (slackBoundX[i] != kInf) {
                constraints.push_back(xi <= bound.upper + slack(0, static_cast<casadi_int>(i)));
            } else {
                constraints.push_back(xi <= bound.upper);
            }
        }
    }
    return constraints;
}

std::vector<casadi::MX> MPCBase::inputConstraints(const casadi::MX& u, const HardBound& hardBound) const {
    std::vector<casadi::MX> constraints;
    for (size_t i = 0; i < hardBound.u.size(); ++i) {
        const Bound& bound = hardBound.u[i];
        casadi::MX ui = u(casadi::Slice(), static_cast<casadi_int>(i));
        if (bound.lower != -kInf) {
            constraints.push_back(ui >= bound.lower);
        }
        if (bound.upper != kInf) {
            constraints.push_back(ui <= bound.upper);
        }
    }
    return constraints;
}

} // namespace npmpc::mpc::problem
